turbopack/crates/turbo-tasks-backend/src/backend/mod.rs

```diff
@@ -0,0 +1,1084 @@
+
+    persisted_task_cache_log: Mutex<ChunkedVec<(Arc<CachedTaskType>, TaskId)>>,
+    task_cache: BiMap<Arc<CachedTaskType>, TaskId>,
+    transient_tasks: DashMap<TaskId, Arc<TransientTask>>,
+
+    persisted_storage_log: Mutex<ChunkedVec<CachedDataUpdate>>,
+    storage: Storage<TaskId, CachedDataItem>,
+
+    /// Number of executing operations + Highest bit is set when snapshot is
+    /// requested. When that bit is set, operations should pause until the
+    /// snapshot is completed. When the bit is set and in progress counter
+    /// reaches zero, `operations_completed_when_snapshot_requested` is
+    /// triggered.
+    in_progress_operations: AtomicUsize,
+
+    snapshot_request: Mutex<SnapshotRequest>,
+    /// Condition Variable that is triggered when `in_progress_operations`
+    /// reaches zero while snapshot is requested. All operations are either
+    /// completed or suspended.
+    operations_suspended: Condvar,
+    /// Condition Variable that is triggered when a snapshot is completed and
+    /// operations can continue.
+    snapshot_completed: Condvar,
+}
+
+impl Default for TurboTasksBackend {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TurboTasksBackend {
+    pub fn new() -> Self {
+        Self {
+            persisted_task_id_factory: IdFactoryWithReuse::new(1, (TRANSIENT_TASK_BIT - 1) as u64),
+            transient_task_id_factory: IdFactoryWithReuse::new(
+                TRANSIENT_TASK_BIT as u64,
+                u32::MAX as u64,
+            ),
+            persisted_task_cache_log: Mutex::new(ChunkedVec::new()),
+            task_cache: BiMap::new(),
+            transient_tasks: DashMap::new(),
+            persisted_storage_log: Mutex::new(ChunkedVec::new()),
+            storage: Storage::new(),
+            in_progress_operations: AtomicUsize::new(0),
+            snapshot_request: Mutex::new(SnapshotRequest::new()),
+            operations_suspended: Condvar::new(),
+            snapshot_completed: Condvar::new(),
+        }
+    }
+
+    fn execute_context<'a>(
+        &'a self,
+        turbo_tasks: &'a dyn TurboTasksBackendApi<Self>,
+    ) -> ExecuteContext<'a> {
+        ExecuteContext::new(self, turbo_tasks)
+    }
+
+    fn suspending_requested(&self) -> bool {
+        (self.in_progress_operations.load(Ordering::Relaxed) & SNAPSHOT_REQUESTED_BIT) != 0
+    }
+
+    fn operation_suspend_point(&self, suspend: impl FnOnce() -> AnyOperation) {
+        if (self.in_progress_operations.load(Ordering::Relaxed) & SNAPSHOT_REQUESTED_BIT) != 0 {
+            let operation = Arc::new(suspend());
+            let mut snapshot_request = self.snapshot_request.lock();
+            if snapshot_request.snapshot_requested {
+                snapshot_request
+                    .suspended_operations
+                    .insert(operation.clone().into());
+                let value = self.in_progress_operations.fetch_sub(1, Ordering::AcqRel) - 1;
```

Comment:
That's not possible because the value must be at least 1 here since this is called during an in progress operation, which incremented the value at the start.

The check at the start doesn't check for the value to be 0 it checks for a different bit.